feat: proposal pre/post test steps (#210)
Agoric/agoric-3-proposals: README.md

- `package.json` specifies what kind of proposal it is in a `agoricProposal` field. If it's a "Software Upgrade Proposal" it also includes additional parameters.
- `use.sh` is the script that will be run in the USE stage of the build
- `test.sh` is the script that will be _included_ in the TEST stage of the build, and run in CI
- `setup-test.sh` is an optional script which can be used to run setup steps _inside_ the container. It runs _before_ the chain starts.
- `teardown-test.sh` is an optional script which can be used to run teardown steps _inside_ the container. It runs _after_ the chain stops.

> - `before-test-run.sh` is an optional script which can be used to run setup steps on the _host_ (like starting a follower). It runs _before_ the container launches. It needs to be placed in the **host** folder.
> - `after-test-run.sh` is an optional script which can be used to run teardown steps on the _host_ (like stopping a follower). It runs _after_ the container exits. It needs to be placed in the **host** folder.

Reviewer: Let's be more explicit about what the `host` folder is for. There has been some confusion: https://github.com/Agoric/agoric-sdk/pull/10947#discussion_r1943969298

```suggestion
- `host` folder is for ___ 
  - `host/before-test-run.sh` is an optional script which can be used to run setup steps on the _host_ (like starting a follower). It runs _before_ the container launches.
  - `host/after-test-run.sh` is an optional script which can be used to run teardown steps on the _host_ (like stopping a follower). It runs _after_ the container exits.
```

While you're at it a change like this would help,
```
 - `test` folder is for files that should be included only in the `test` image
```

This is not blocking feedback. If something is delayed by this PR the docs improvements can come later.

Author: Added

- `package.json` specifies what kind of proposal it is in a `agoricProposal` field. If it's a "Software Upgrade Proposal" it also includes additional parameters.
- `use.sh` is the script that will be run in the USE stage of the build
- `test.sh` is the script that will be _included_ in the TEST stage of the build, and run in CI

> - `prepare-test.sh` is an optional script which can be used to run setup steps _inside_ the container

Reviewer: in this PR would you [revise the name](https://github.com/Agoric/agoric-3-proposals/pull/209#discussion_r1904570229)?  Please clarify that it runs within the test image before agd starts. `test-before-agd.sh` would be clear.

This readme must also mention the two new hooks.

Author: I think the hook name should not have any mention of `agd`. It should just be an indication of at which point in the pipeline this hook fits and `prepare-test.sh` I think conveys that. We could rename it to `pre-test.sh` as well but we then have to distinguish on hook names running outside and inside the container. The current PR uses `pre-test.sh` name which run on the host.
What do you suggest?

Reviewer: > It should just be an indication of at which point in the pipeline this hook fits

Agree

> and prepare-test.sh I think conveys that.

Disagree. `prepare.sh` works very differently,
> PREPARE For upgrade proposals: submits the proposal, votes on it, runs to halt for the next stage

`prepare-test.sh` sounds like it would be a PREPARE step but for tests. However it runs in the same image as `test.sh`. It's just like `test.sh` but runs before agd starts.

I'll argue firmly that we need a different name than `prepare-test.sh`. As you point out `pre` and `post` are taken. I think those should be more explicit that they don't run within the image: `before-test-run.sh` and `after-test-run.sh`.  ("Before" and "after" more closely match testing hook names like Ava has.)

That would leave `pre-test.sh` but that could still be easily confused. I would suggest `setup-test.sh` or `test-before-chain.sh`. 

If we should discuss more let's do it synchronously or in a call.

Author: I agree with `before-test-run.sh` and `after-test-run.sh` for host and `setup-test.sh` (and possibly `teardown-test.sh`) for image so we can lock them if you agree as well.

Reviewer: 🤝
